### PR TITLE
Fixmatcher Modified resolves #6366

### LINKED ISSUE #6366: DescriptorMatcher::checkMasks fails on Nx1 masks
### Please state the information for your system
- OpenCV version: 3.x
- Host OS: Linux (Ubuntu 14.04)

### In which part of the OpenCV library you got the issue?
- feature2d
- matcher

### Expected behaviour
Running knnMatch or radiusMatch with a Nx1 mask pass on the assertion inside DescriptorMatcher::checkMasks method.

### Actual behaviour
Running fails with the following error
> 
matchers.cpp:617: error: (-215) masks[i].rows == queryDescriptorsCount && (masks[i].cols == rows || masks[i].cols == rows) && masks[i].type() == CV_8UC1 in function checkMasks

### Additional description
The assertion fails becaus the second condition of the assertion only compares the number of columns of the masks.

### Code example to reproduce the issue / Steps to reproduce the issue
```c++
#include <iostream>

#include <string>
#include <vector>

#include <opencv2/highgui.hpp>
#include <opencv2/videoio.hpp>

#include <opencv2/features2d.hpp>
#include <opencv2/imgproc.hpp>
#include <opencv2/video/video.hpp>

int main(int argc, char** argv) {

    std::string folder("");

    std::vector<cv::Mat> images;
    images.push_back(cv::imread("/home/anishitani/workspace/visionws/src/tutorial/data/fountain/0000.jpg",cv::IMREAD_GRAYSCALE));
    images.push_back(cv::imread("/home/anishitani/workspace/visionws/src/tutorial/data/fountain/0001.jpg",cv::IMREAD_GRAYSCALE));

    /** Feature detection init **/
    std::vector<std::vector<cv::KeyPoint> > fkpts(2);

    cv::Ptr<cv::FeatureDetector> detector = cv::FastFeatureDetector::create();
    detector->detect(images[0],fkpts[0]);
    detector->detect(images[1],fkpts[1]);
    /** Feature detection end **/

    /** Feature matching init **/
    std::vector<std::vector<cv::Point2f> > fpts(2);
    cv::KeyPoint::convert(fkpts[0],fpts[0]);
    cv::KeyPoint::convert(fkpts[1],fpts[1]);
    // Calculate the optical flow field:
    // how each left_point moved across the 2 images
    std::vector<uchar> vstatus;
    std::vector<float> verror;
    cv::calcOpticalFlowPyrLK(images[0], images[1], fpts[0], fpts[1], vstatus, verror);
    /** Feature matching end **/

    cv::Mat out;
    cv::cvtColor(images[0],out,cv::COLOR_GRAY2BGR);
    std::vector<cv::Point2f> flowpts;
    for(int i=0 ; i<fpts[0].size() ; i++){
        if(vstatus[i] && verror[i]<12.0){
            flowpts.push_back(fpts[1][i]);
        } else {
            flowpts.push_back(cv::Point2f());
            vstatus[i] = 0;
        }
    }

    cv::BFMatcher matcher;
    std::vector<std::vector<cv::DMatch> > nn;
    cv::Mat mask;
    cv::Mat(vstatus).convertTo(mask,CV_8UC1);

    matcher.radiusMatch( cv::Mat(fpts[1]).reshape(1,fpts[1].size()),
            cv::Mat(flowpts).reshape(1,flowpts.size()),
            nn,2.0f, mask);

    cv::namedWindow("features",cv::WINDOW_KEEPRATIO);
    cv::imshow("features",out);

    for(char k = cv::waitKey() ; k!='q' ; k = cv::waitKey());

    return 0;
}

```

### PR DESCRIPTION
**Pull request by souch55**
#6366

### CODE REVIEW

File: modules/features2d/src/matchers.cpp [hunk 1]
```diff
@@ -613,7 +613,7 @@ void DescriptorMatcher::checkMasks( InputArrayOfArrays _masks, int queryDescript
             {
                 int rows = trainDescCollection[i].empty() ? utrainDescCollection[i].rows : trainDescCollection[i].rows;
                     CV_Assert( masks[i].rows == queryDescriptorsCount &&
-                        (masks[i].cols == rows || masks[i].cols == rows) &&
+                        (masks[i].rows == rows || masks[i].cols == rows) &&
```
Comment (terfendail) [inline modules/features2d/src/matchers.cpp]: masks[i].rows is already checked at line 615 so only masks[i].cols should be checked here. masks[i] should be a matrix with query descriptors count rows and train descriptors count columns(in accordance with [parameters description](http://docs.opencv.org/3.1.0/db/d39/classcv_1_1DescriptorMatcher.html#a8737756123c4aba88ae5fc15982e0dee) for DescriptorMatcher::Match)

**Comment by souch55** [comment]
Hi terf thanks for yours comments but when tested with the app and yours suggested solution  the bug re-occur #https://github.com/opencv/opencv/issues/6366,what ever the current fixation has been done also suggested by catee #https://github.com/opencv/opencv/pull/7103, is correct only,coz assertion passes with it.

**Comment by mshabunin** [comment]
@souch55 , you don't need to create new pull request each time you update the code, just push to the same branch in your repository and the pull request will get updated.

As to the issue, looks like the bug is not valid actually.

**Comment by souch55** [comment]
Hi mushabuin thanks for yours comments,yes i will use push to same branch for modification and then pull it,about the issue here,mask width is checked twice,which i have tried to fix here in my  pull request.

**Comment by terfendail** [comment]
As far as I could understand the mask parameter description the assertion should NOT pass for the Nx1 mask matrix. Regarding the #6366 mask parameter should be NxM matrix with rows filled with all 0 for each 0 of Nx1 mask matrix and with all 1 otherwise. Also if the Nx1 mask will pass the assertion it could lead to out of bound reading during call to DescriptorMatcher::isPossibleMatch that expect it to be NxM

**Comment by souch55** [comment]
Hi terfendail thanks for yours comments on this issue which i am trying to understand and shout out,this line as in master source code line 616 "(masks[i].cols == rows || masks[i].cols == rows) ",is wrong right,its not meaning full code here ,so what you think is it need not to be corrected?

**Comment by mshabunin** [comment]
@souch55 , you just need to remove second `masks[i].cols == rows` expression.

**Comment by terfendail** [comment]
Changes looks fine for me.
@souch55 Could you please squash both commits into one?

**Comment by alalek** [comment]
> "[Commits 4](https://github.com/opencv/opencv/pull/7114/commits)"

Something goes wrong. Please don't use merge or pull commands.

Use "git rebase -i HEAD~4" to squash commits.
Push updated changes on GitHub via command like this: "git push -f origin Fixmatchar" (-f is required)

**Comment by souch55** [comment]
Hi alalek thanks for urs comments, i did both merge and pull and then rebase-squash ,but its seems like something as went wrong ,please find my fixation for the issue#6366 thing here in this new request #7130

**Pull request closed without merging**